lib/makara/proxy.rb

```diff
@@ -162,6 +166,43 @@ def any_connection
       end
     end
 
+    MAX_RETRY_COUNT = 10
+
+    def _execute_with_connection_and_retry_exceptions(retry_exceptions, args, block, method_name)
+      begin
+        should_retry = false
+        potentially_stale_connection = nil
+
+        appropriate_connection(method_name, args) do |con|
+          con.send(method_name, *args, &block)
+          potentially_stale_connection = con
+        end
+
+      rescue Exception => actual_exception
+        retry_attempts = retry_exceptions.inject({}) { |memo, retry_exception| memo[retry_exception['name']] = 0; memo }
+        retry_exceptions.each do |retry_exception|
+          if actual_exception.class.to_s == retry_exception['name']
+            potentially_stale_connection.disconnect!
+            sleep retry_exception['time_between_retries_in_seconds'] || 0.1
```

Comment:
I wonder if a bit of randomization here would be useful to avoid a load spike after the instance comes up.